File: src/main/scala/com/yogpc/qp/integration/jei/WorkBenchRecipeCategory.scala
package com.yogpc.qp.integration.jei

import com.yogpc.qp.integration.jei.WorkBenchRecipeCategory._
import com.yogpc.qp.{QuarryPlus, QuarryPlusI}
import mezz.jei.api.IGuiHelper
import mezz.jei.api.gui.IDrawableAnimated.StartDirection
import mezz.jei.api.gui.{IDrawable, IRecipeLayout}
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.recipe.IRecipeCategory
import net.minecraft.client.Minecraft
import net.minecraft.util.ResourceLocation

class WorkBenchRecipeCategory(guiHelper: IGuiHelper) extends IRecipeCategory[WorkBenchRecipeWrapper] {
  //4, 13 => 175, 94

  val bar = guiHelper.createDrawable(backGround, xOff, 87, 160, 4)
  val animateBar = guiHelper.createAnimatedDrawable(bar, 300, StartDirection.LEFT, false)

  override val getBackground: IDrawable = guiHelper.createDrawable(backGround, xOff, yOff, 167, 86)

  override def setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: WorkBenchRecipeWrapper, ingredients: IIngredients): Unit = {
    val guiItemStack = recipeLayout.getItemStacks
    //7, 17 -- 7, 89

    val x0 = 3
    for (i <- 0 until recipeWrapper.recipeSize) {
      val row = i / 9
      val columns = i % 9
      guiItemStack.init(i, true, x0 + o * columns - xOff, x0 + o * row - yOff)
    }
    guiItemStack.init(recipeWrapper.recipeSize, false, x0 - xOff, x0 + 64 - yOff)
    guiItemStack.set(ingredients)
  }

  override def getTitle: String = QuarryPlusI.blockWorkbench.getLocalizedName

  override val getUid: String = UID

  override def drawExtras(minecraft: Minecraft): Unit = {
    animateBar.draw(minecraft, 4, 60)
  }

  override def getModName: String = QuarryPlus.Mod_Name
}

object WorkBenchRecipeCategory {
  final val UID = "quarryplus.workbenchplus"
  val backGround = new ResourceLocation(QuarryPlus.modID, "textures/gui/workbench_jei2.png")
  final val xOff = 0
  final val yOff = 0
  final val o = 18
}

File: src/main/scala/com/yogpc/qp/integration/jei/WorkBenchRecipeCategory.java
package com.yogpc.qp.integration.jei;

import java.util.Collections;

import com.mojang.blaze3d.vertex.PoseStack;
import com.yogpc.qp.Holder;
import com.yogpc.qp.QuarryPlus;
import com.yogpc.qp.machines.PowerTile;
import com.yogpc.qp.machines.workbench.IngredientList;
import com.yogpc.qp.machines.workbench.WorkbenchRecipe;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.gui.builder.IRecipeLayoutBuilder;
import mezz.jei.api.gui.drawable.IDrawable;
import mezz.jei.api.gui.drawable.IDrawableAnimated;
import mezz.jei.api.gui.drawable.IDrawableStatic;
import mezz.jei.api.gui.ingredient.IRecipeSlotsView;
import mezz.jei.api.helpers.IGuiHelper;
import mezz.jei.api.recipe.IFocusGroup;
import mezz.jei.api.recipe.RecipeIngredientRole;
import mezz.jei.api.recipe.RecipeType;
import mezz.jei.api.recipe.category.IRecipeCategory;
import net.minecraft.client.Minecraft;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;

class WorkBenchRecipeCategory implements IRecipeCategory<WorkbenchRecipe> {
    public static final RecipeType<WorkbenchRecipe> RECIPE_TYPE = RecipeType.create(QuarryPlus.modID, "jei_workbenchplus", WorkbenchRecipe.class);
    private static final ResourceLocation backGround = new ResourceLocation(QuarryPlus.modID, "textures/gui/workbench_jei2.png");
    private static final int xOff = 0;
    private static final int yOff = 0;
    private final IGuiHelper helper;
    private final IDrawableAnimated animateBar;

    WorkBenchRecipeCategory(IGuiHelper helper) {
        this.helper = helper;
        IDrawableStatic bar = helper.createDrawable(backGround, xOff, 87, 160, 4);
        this.animateBar = helper.createAnimatedDrawable(bar, 300, IDrawableAnimated.StartDirection.LEFT, false);
    }

    @Override
    public RecipeType<WorkbenchRecipe> getRecipeType() {
        return RECIPE_TYPE;
    }

    @Override
    public Component getTitle() {
        return Holder.BLOCK_WORKBENCH.getName();
    }

    @Override
    public IDrawable getBackground() {
        return helper.createDrawable(backGround, xOff, yOff, 167, 86);
    }

    @Override
    public IDrawable getIcon() {
        return helper.createDrawableIngredient(VanillaTypes.ITEM_STACK, new ItemStack(Holder.BLOCK_WORKBENCH));
    }

    @Override
    public void draw(WorkbenchRecipe recipe, IRecipeSlotsView recipeSlotsView, PoseStack stack, double mouseX, double mouseY) {
        animateBar.draw(stack, 4, 60);
        Minecraft.getInstance().font.draw(stack, (recipe.getRequiredEnergy() / PowerTile.ONE_FE) + "MJ", 36 - xOff, 70 - yOff, 0x404040);
        // Enchantment copy
        // Minecraft.getInstance().font.drawString(matrixStack, (recipe.energy.toDouble / APowerTile.MJToMicroMJ).toString + "MJ", 36 - xOff, 67 - yOff, 0x404040)
        // Minecraft.getInstance().font.drawString(matrixStack, "Keeps enchantments", 36 - xOff, 77 - yOff, 0x404040)
    }

    @Override
    public void setRecipe(IRecipeLayoutBuilder builder, WorkbenchRecipe recipe, IFocusGroup focuses) {
        var input = recipe.inputs().stream()
            .map(IngredientList::stackList)
            .toList();
        var output = Collections.singletonList(recipe.output);

        int x0 = 4;
        final int o = 18;
        for (int i = 0; i < recipe.inputs().size(); i++) {
            int xIndex = i % 9;
            int yIndex = i / 9;
            var slotInput = input.get(i);
            builder.addSlot(RecipeIngredientRole.INPUT, x0 + o * xIndex - xOff, x0 + o * yIndex - yOff)
                .addItemStacks(slotInput);
        }
        builder.addSlot(RecipeIngredientRole.OUTPUT, x0 - xOff, x0 + 64 - yOff)
            .addItemStacks(output);
    }
}
